[김은혜] Step1 엔티티 매핑 PR 요청

eunveloper opened the pull request:
안녕하세요! 2단계 JPA 시작의 PR 올립니다!
1단계에서도 정말 많이 배웠습니다. 2단계에서도 많이 배우길 기대하고 있어요 :)

소중한 피드백 부탁드려요! 많이 보고 많이 배우겠습니다.
감사합니다!

eunveloper committed:
docs: README.md 작성

- 요구사항 작성
- 관련 sql문 작성

eunveloper committed:
feat: Entity 객체 sql 맞춰 구성

- 패키지 구조 세분화
- Entity 객체 JPA 어노테이션 적용
- 공통 시간값 관리하는 BaseTime 객체 생성

eunveloper committed:
feat: 테스트코드 작성

- AnswerTest 테스트 작성
- QuestionTest 테스트 작성
- UserTest 테스트 작성

devyonghee commented on README.md:
```diff
@@ -0,0 +1,64 @@
+# 1단계 - 엔티티 매핑
+
+## 요구 사항
+
+QnA 서비스를 만들어가면서 JPA로 실제 도메인 모델을 어떻게 구성하고 객체와 테이블을 어떻게 매핑해야 하는지 알아본다.
+* 아래의 DDL(Data Definition Language)을 보고 유추하여 엔티티 클래스와 리포지토리 클래스를 작성해 본다.
+* @DataJpaTest를 사용하여 학습 테스트를 해 본다.
+
```
요구사항 명확하게 잘 정리해주신 것 같아요 👍👍

devyonghee commented on src/main/java/qna/Application.java:
```diff
@@ -2,7 +2,9 @@
 import org.springframework.boot.SpringApplication;
 import org.springframework.boot.autoconfigure.SpringBootApplication;
+import org.springframework.data.jpa.repository.config.EnableJpaAuditing;
 
+@EnableJpaAuditing(modifyOnCreate = false)
```
`modifyOnCreate` 옵션 주신 부분이 인상적입니다 👍👍👍

devyonghee commented on src/main/java/qna/domain/Answer.java:
```diff
@@ -3,15 +3,25 @@
-public class Answer {
+@Entity
+public class Answer extends BaseTime {
+
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
     private Long id;
+
     private Long writerId;
     private Long questionId;
     private String contents;
+
```
`contents` 의 생성되는 데이터 컬럼 타입이 `varchar` 라서 요구사항에 명시되어 있는 타입과 다른데요 🤔
다음 애너테이션을 참고해주시고 확인 한번만 부탁드리겠습니다.🙇‍♂️ 
[@Lob (Large Object)](https://kogle.tistory.com/250)

eunveloper replied:
JPA를 실무에서 많이 써봤는데, DDL 생성에 초점을 둔적은 없었어서
이렇게 세부적으로 어노테이션이 다 있는지 몰랐네요! 반영하겠습니다 😆

devyonghee commented on src/main/java/qna/domain/BaseTime.java:
```diff
@@ -0,0 +1,39 @@
+@EntityListeners(AuditingEntityListener.class)
+public class BaseTime {
+
+    @CreatedDate
+    @Column(nullable = false)
+    private LocalDateTime createdAt;
+
+    @LastModifiedDate
+    private LocalDateTime updatedAt;
```
이 필드들에 `setter` 가 필요한지 확인 해보시면 좋을 것 같아요 😆
데이터는 괜찮겠지만 `setter` 로 인해 혼란이 생기지 않을까 생각되는데요. 

기본적으로 `setter` 가 없어도 정상적으로 동작하는 알고 있긴 하지만 
은혜님에게 자세한 피드백을 드리고자 데이터들이 어떻게 세팅이 되는지 살펴봤는데요.

`org.springframework.data.auditing.AuditingHandlerSupport#touchDate` 에서 데이터 세팅을 시작하여  

`org.springframework.data.mapping.model.ClassGeneratingPropertyAccessorFactory.PropertyAccessorClassGenerator#visitSetProperty`  필드에 직접 접근하여 값을 세팅할 수 있도록 새로운 메서드를 선언하여 

`org.springframework.data.mapping.model.ClassGeneratingPropertyAccessorFactory#createAccessorClass` 에서 `Reflection` 으로 새로운 클래스를 생성하여 이용하는 모습이 보입니다.

그렇기 때문에 `setter` 가 없어도 잘 동작할 것 같은데,
혹시 혼란이 생기지 않도록 `setter` 를 제거하는 방향에 대해서 어떻게 생각하시나요??

eunveloper replied:
너무 관행적으로 setter getter를 작성해버렸네요 ㅠㅠ
더군다나 이 클래스는 JpaAudting 기능으로 자동으로 생성 수정시 마다 데이터를 적용하기때문에
개발자가 직접 setter할 일은 더더욱 없어 보여요!

세팅과정까지 자세히 설명해주셔서 감사드려요! 😇

devyonghee commented on src/main/java/qna/domain/DeleteHistory.java:
```diff
@@ -1,15 +1,28 @@
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
     private Long id;
+
     private ContentType contentType;
```
`enum` 타입은 기본 설정이 `ORDINAL` 로 설정되어 있기 때문에 별도의 설정이 없으면 `integer` 타입으로 사용하게 되는데요. 
알고 계시겠지만 `@Enumerated` 을 활용해보시는 것을 추천드릴게요!! 😆

devyonghee commented on src/main/java/qna/domain/Question.java:
```diff
@@ -1,12 +1,25 @@
+
+    @Column(nullable = false)
     private String title;
+
```
길이 설정이 다른 것 같아요 확인 한번만 부탁드리겠습니다 🙇‍♂️🙏

devyonghee commented on src/main/java/qna/domain/Question.java:
```diff
@@ -1,12 +1,25 @@
+    @Column(nullable = false)
     private String title;
+
     private String contents;
```
위에서 코멘트 드려서 아시겠지만 요구사항에 있는 ddl 의 데이터 타입이 다른 것 같아요.
확인 한번만 부탁드리겠습니다.🙇‍♂️

devyonghee commented on src/test/java/qna/domain/AnswerTest.java:
```diff
@@ -1,6 +1,56 @@
+    @Autowired
+    private AnswerRepository answerRepository;
+
+    @Test
+    @DisplayName("Answer 객체를 저장하면 Id가 자동생성 되어 Not Null 이다.")
+    void save() {
+        assertThat(A1.getId()).isNotNull();
+        assertThat(answerRepository.save(A1).getId()).isNotNull();
+        assertThat(answerRepository.save(A2).getId()).isNotNull();
+    }
```
검증을 여러개 수행하고 싶다면 
Junit 의 `assertAll` 을 사용하는 방법은 어떻게 생각하시나요? 🤔
이 메서드를 사용하면 중간에 예외가 발생해도 그 이후의 검증 로직도 확인할 수 있어서 좋을 것 같아요 😆
참고해보시면 좋을 것 같은 자료 제공해드리겠습니다!! [assertAll](https://gracelove91.tistory.com/108)

eunveloper replied:
여러개의 검증을 assertAll을 통해서! 기억하고 반영해보고 갈게요 ㅎㅎ

devyonghee commented on src/test/java/qna/domain/AnswerTest.java:
```diff
@@ -1,6 +1,56 @@
+        answerRepository.save(A1);
+        assertThat(answerRepository.findByWriterId(1L).isPresent()).isTrue();
+        assertThat(answerRepository.findByWriterId(1L).get().getContents()).isEqualTo(A1.getContents());
+        assertThat(answerRepository.findByWriterId(10L).isPresent()).isFalse();
```
assertj 를 활용하시면 `Optional` 에 대한 검증과 메서드 체이닝을 통해 검증이 가능한데요.
다음과 같이 검증도 가능하니 참고만 해주세요 🙏

```suggestion
        assertThat(questionRepository.findByWriterId(1L)).isPresent()
                .get()
                .extracting(Question::getContents)
                .isEqualTo(Q1.getContents());
        assertThat(questionRepository.findByWriterId(10L)).isEmpty();
```

devyonghee commented on src/test/java/qna/domain/AnswerTest.java:
```diff
@@ -1,6 +1,56 @@
+    @Test
+    @DisplayName("Answer 객체를 저장하면 Id가 자동생성 되어 Not Null 이다.")
+    void save() {
+        assertThat(A1.getId()).isNotNull();
```
이 메소드만 단독으로 실행하면 테스트가 깨지게 됩니다 😅
다른 테스트에 대해 영향을 받고 있어서 전체 테스트가 통과 된건지 확인 한번만 부탁드리겠습니다 🙇‍♂️

eunveloper replied:
save() 전에는 id가 Null이고 save() 후엔 id가 Notnull인걸 확인하려고 했는데
메소드가 반대로 구성되었네요 ㅠㅠ 이런실수를.... 수정하겠습니다~!

devyonghee commented on src/test/java/qna/domain/QuestionTest.java:
```diff
@@ -1,6 +1,54 @@
+    @DisplayName("Question 객체를 수정하면 수정된 데이터와 일치해야 하고 업데이트 날짜가 Not Null 이다.")
+    void update() {
+        Question question = new Question("title1", "contents1").writeBy(UserTest.JAVAJIGI);
+        Question actual = questionRepository.save(question);
+        assertThat(actual.getUpdatedAt()).isNull();
+
```
현재 테스트에서 2가지를 검증하고 있는 것은 아닌가 라는 생각이 드는데요 🤔
저장했을 때 `updatedAt` 이 비어있는 것을 검증하는 것은 `save` 테스트에서 이뤄져야 하는 것 이 아닌가 생각됩니다. 
혹시 이러한 부분에 대해서 어떻게 생각하시나요 🤔🤔🤔

테스트를 BDD 의 행동으로 구분해서 작성해보시면 도움이 되실 것 같아요 😆
제가 고민해보고 참고한 자료 남겨드리도록 하겠습니다!! [BDD 스타일](https://effortguy.tistory.com/145)

eunveloper replied:
용희님 말씀이 맞는거 같아요!
updatedAt 이 비어있는건 insert(save) 라는 행위(Behavior) 로 인해서 확인이 되는 결과이기 때문에
업데이트를 하는 로직의 테스트 코드보단 저장의 로직에 관한 테스트 코드가 적절할 것 같네요! 🤣

devyonghee reviewed (changes requested):
안녕하세요 김은혜님!! 👋
두 번째 JPA 미션을 함께 하게 된 한용희라고 합니다!🙇‍♂️🙇‍♂️🙇‍♂️
엔티티 매핑도 잘해주시고 테스트도 꼼꼼하게 진행해주셨네요!! 👍😆😆

그런데 간혹 요구사항에 있는 ddl 대로 매핑되지 않는 부분들이 보이고
단독으로 실행할 경우 깨지는 테스트가 존재해서 확인 한번만 부탁드리겠습니다. 🙇‍♂️🙏
그 외에 추가적으로 소소한 코멘트들도 남겨드렸는데 의견 부탁드리겠습니다.

p.s

> 1단계에서도 정말 많이 배웠습니다. 2단계에서도 많이 배우길 기대하고 있어요 :)

굉장히 부담스러운 말씀을 해주셔서 어깨가 굉장히 무거워졌습니다.. 😓🫠

eunveloper committed:
refactor: 코드리뷰 피드백 반영

- updatedAt의 테스트 시점 변경 (update -> save)
- 초기 save() id 값 테스트 변경
- 데이터타입 clob 어노테이션 반영
- 테이터타입 enum @Enumerated 반영
- 단독 및 통합테스트를 위해 static 객체 불필요한 부분 걷어내기
- 문자열 데이터 길이값 반영

eunveloper committed:
refactor: 코드리뷰 피드백 반영

- 불필요 setter 메소드 제거
- 다중 테스트 메소드에 assertAll 활용
- Optional 객체 테스트 검증방식 변경

eunveloper committed:
refactor: 코드리뷰 피드백 반영

- Optional 객체 테스트 검증방식 변경

devyonghee commented on src/test/java/qna/domain/QuestionTest.java:
```diff
@@ -1,6 +1,62 @@
+    @Test
+    @DisplayName("Question 객체를 저장하면 Id가 자동생성 되어 Not Null 이다.")
+    void save() {
+        Question question = new Question("title1", "contents1").writeBy(UserTest.JAVAJIGI);
+        assertThat(question.getId()).isNull();
```
이 테스트는 도메인의 테스트에서 진행되어야 하는 사항이 아닌가 라는 생각이 드는데
이 부분에 대해 어떻게 생각하시나요?🤔

devyonghee reviewed (approved):
안녕하세요 은혜님 
늦어서 죄송합니다 🙇‍♂️🙇‍♂️🙇‍♂️
피드백을 수용해주시고 수정해주셔서 감사합니다 👍🙏

엔티티 매핑 잘해주신 것 같아요 👍👍👍
이번 단계는 여기서 마무리하고 머지하도록 하겠습니다.
그럼 다음 단계도 파이팅입니다!! 🚀💪
